### DISCUSSION TITLE
On depreciation and new unstable features - Defining a depreciation strategy

### DISCUSSION THREAD

**Discussion by kienerj** [rdkit/rdkit#6758] [Development]
Nowadays I'm really more and more into the Software Engineering things than pure Cheminformatics. From my experience with the RDKit it is really missing a proper depreciation strategy to get rid of old crutch and be able to improve things. The crutch is also driven by new features that have issues and get improved over time but making the new versions only available via "version" or "feature" arguments to the method call or adding a new method with the versioning in the method name.

Currently the balance is way too much on the backwards compatibility-side.

As an example I want to use the RegistrationHash added in as far as I remember in 2022.9. This initial version has a severe issue with the tautomer hash removing double bond stereo, see [here](https://github.com/rdkit/rdkit/discussions/6318). In my opinion this makes it pretty much unusable. In 2023.3 this is fixed but only if you set the according feature flag:

    RegistrationHash.GetMolLayers(m, enable_tautomer_hash_v2=True)

Anyone wanting to use the in my opinion only actual working version of RegistrationHash must set this flag "for eternity".  

It's hard to keep track of all the flags needed to get the intended version of the methods you are calling. In a final application, one must comment the code section to ensure nobody tries to remove the flag(s) and to explain why they need to be there.

**My Proposal:**

- Define a depreciation strategy
  -  Example: mark method/class/module as deprecated (in C++ and python) and give 3 years / major versions for users to apply till the code is removed

- Mark all new features as unstable
  - If issues are found there can be breaking changes to fix the issues without having to resort to feature flags
  - I suggest keeping features unstable for 1 year / major version (gets a bit tricky if new features are being added in patch versions)
 
Very interested in the communities thought on this.

**Comment by greglandrum** [comment]
Thanks for starting this discussion @kienerj 
There are multiple topics here that are all worth talking about.

# Changing default behavior of functions

I agree that, whenever possible, the RDKit should do the "right" ("best"?) thing by default. This may mean making changes to the code which result in "client" code (i.e. code which uses the RDKit) generating different results. We have done this in the past and will continue to do so; such changes get tagged with "Changes behavior" here in GitHub and are documented in the "Backwards incompatible changes" section of the ReleaseNotes.md document. I think this approach works and I don't recall complaints from the community about it.

In the specific case of the tautomer hash and `GetMolLayers()`: I didn't make v2 of the tautomer hash the default when we addeed it to the RDKit because the approach is scientifically new (in the sense that I wasn't implementing a published method) and I wasn't confident that it was good enough to be the default. Once we have more experience with it (maybe by the 2024.03 release), I completely agree that the v2 hash should become the default.

# Deprecating/removing things

You're right that we should be more explicit about what the policy is. At the moment the release notes have sections for "Deprecated code (to be removed in a future release)" and "Code removed in this release", but we could make the list of deprecated items (and when they will be removed) more visible.

@PatrickPenner did some work at the hackathon to make deprecations more visible. Hopefully that turns into a PR at some point.

# Marking new stuff as unstable

This is a good idea for features/additions where we aren't sure about the API (I don't think I'd want to automatically mark everything new as "unstable" though). There are already a few parts of the code where we mention that the API/results may change from release to release, but these are not easy to discover.


# Making things more discoverable

I'm curious what others think the best way to make deprecations and "unstable" features visible is. Clearly we can (and will) continue to put things in the release notes (we can add as section to those for the unstable features), but it seems like it would be useful to have a place where this information is easily gathered in one place. Maybe in the GitHub wiki? Or perhaps a separate document in the repo itself?

**Comment by kienerj** [comment]
For deprecations in Python there for now is a package that emulates the Java behavior of using `@deprecated` to mark classes or functions as deprecated and be able for IDEs to give warning about using such functions. There is also an official [PEP](https://peps.python.org/pep-0702/) for this but only for Python 3.13.

So I think it will take a while to make this more easily available for developers (in python) directly without looking at some sort of documentation.  maybe something similar exists in C++? (indeed `[[deprecated]]` but no idea if it is already used or useful at all)

**Reply by PatrickPenner** [reply]
Hi @kienerj, I took a look at mechanisms of deprecation during the hackathon. There is now a pull request here: https://github.com/rdkit/rdkit/pull/6767

In that pull request I use [[deprecated( string-literal)]] to generate compile-time warnings when you use deprecated functions from cpp. This is standard enough for cpp but doesn't help us with python.

In the few functions I deprecate in that pull request I also add a runtime warning using the cpp logging that will show up when a deprecated function is used on the python side. Runtime means that this will not show up in the static code analysis done by IDEs, but it should be reasonably effective at warning the programmer.

There are already pure python deprecation warnings such as here: https://github.com/rdkit/rdkit/blob/32655f5365e363ce13bd6b28e2e9e2544f8680bd/rdkit/Chem/MCS.py#L6
These are also runtime warnings.

Personally, I think runtime warnings should be enough for now. If and when the rdkit is fine with using Python 3.13 features it would be worth it to reconsider attributes on the python side. I would be very hesitant to add a package dependency for static deprecation warnings.

**Reply by kienerj** [reply]
Fully agree to not add an extra package just for depreciation in python.
